### PR TITLE
Windows CI build

### PR DESCRIPTION
**Pull request by atsju**
I'm proud to announce this implements CI to build Windows version of DFTFringe. 
First build will take over 1h because of OpenCV build but every next build will use cache I setup and take under 10 minutes. Cache lives for 7 days.

You will notice I moved openCV 3.4.12 to 4.6.0. I think it is good practice to keep up to date to limit work and help maintenance. And I have not been able to build 3.4.12 anyway so there is no real question here. 4.7.0 breaks build but I might fix later.
I also made relative path for all dependencies in .pro. Maybe we have an other solution but this was easiest for me to make the CI work. Sorry Dale, you will either need to accommodate the path I propose or we set so env var or something to accommodate everybody. I'm open to discussion (chat or call).

I don't know why lapack and blas are .a instead .dll on my build, maybe we can fix that later.

You will be able to download a zip of build result exe and dll in action tab under artifacts. It's named `DFTFringe-build-artifact`. here is an example of the build at the moment of this PR: https://github.com/atsju/DFTFringe/actions/runs/4246078248
Once mergged, actions will run in DFTFringe organisation instead my atsju organisation.
Artifacts live for 90 days. It does not replace releases but will help Dale and others test the proposals without rebuilding himself.

@githubdoe If you write a simple but complete document about creating the installer I could add the step to the build so that the artifact is an actual installer instead the exe that won't work on it's own on a clean computer.

I keep this PR draft for now because I would like the other PRs to be merged before so I can rebase the branch properly.
I also expect to have some discussion here. 

This has already been a lot of work but I have a some improvements in mind. Maybe I should list them is an issue ?

Notify @githubdoe @gr5 @brave-ulysses

### CODE REVIEW

**Comment by gr5** [comment]
wow.

**Comment by githubdoe** [comment]
I'm a windows guy and almost never use the command line.  The installer I use is called NSIS from Nullsoft.
Once opened I click on the "Installer based on .zip file" option.

First I move all the files that are needed into one directory. It includes the following directories.
             ColorMaps
              imageformats
              platforms
              res
and all the .dll files and the ReversionHistory.html file.

Then press the "Installer based on  .zip file" option on NSIS.  Tell it where the zip file is.  Tell it what the install path will be "C:\DFTFringe..."  Tell it what to name the exe it will create.
Press Generate button.
Go to github and create a release and upload that .exe file just created as well as copy and paste the latest revision history part of the revisionHistory.html file.

Announce the release on the group and provide that same section of the revision history stating the updates.

**Comment by gr5** [comment]
I love NSIS.  I've written quite a few installers with quite a few complexities in them in that NSIS scripting language.

**Comment by atsju** [comment]
Thank you Dale for the instructions. 
In this particular case I don't see the added value of NSIS. If my understanding is correct, without customization it just extracts the zip you created manually. I will try to create a complete zip without the NSIS part later today so that the generated artifact is the equivalent of your install folder. If NSIS is needed for a reason I don't know, I should be able to add it.

Do you have any feedback about my commit ? 
What about the path in .pro ? Could you accommodate this change or someone has an other idea of how to do it ?

**Comment by githubdoe** [comment]
The benefit is for the Windows user who does not know computers very well.  They don't have to have or understand Zip.  They just have to press Install and everything else is taken care of for them.    In my opinion that is a must.   That fact that it does not do much more than unzipping is not relevant.  It is an installer that more users are use to.

**Comment by githubdoe** [comment]
I created a new directory on my windows machine to try out the current build from master.  I was surprised that it did not build and stopped with this error that does not tell me much.  This was using Qt creator as the IDE and it's build buttons.  So master may be messed up.  I don't really have much time to discover the cause.  So if anyone can give it a try.

:-1: error: [Makefile.Release:844: release/DFTFringe_res.o] Error 1

**Comment by githubdoe** [comment]
Well I'm even more confused than usual.  I had not select the right project in Creator.  The current version builds just fine.  Sorry.

**Comment by atsju** [comment]
Hi Dale, 
thank you for feedback. I think the zip is OK for CI then, as only (very) advanced users will use it. We will keep the installer for official releases.

What about relative path in .pro ? 
Can 2 other PR be merged ? (I don't wan to hurry you, just to understand if there is a masked process behind merging or a lack of time to read/validate the PRs on your side)

------------

I added necessary files to zip archive. 
Maybe you can download the zip and check if everything is fine on your side.

**Click on details on any of the `build-windows`**

image

**Then summary** 

image

**Then download the artifact**

image

**Comment by githubdoe** [comment]
Sorry,  I'm pretty busy with other tasks.  It will take some time for me to check it out.

**Comment by githubdoe** [comment]
I don't want to mess up by current build area on my PC.   So what I want to do is download this (branch?, whatever, I can't find a branch.)  as a zip file and then try to build it using my  Qt configuration.   But I can't figure out where to find the zip if there is one.  If no zip is there an alternative without using git?  Worst case I guess I somehow can use git but I don't know how to do that using a brand new local repro.

**Comment by atsju** [comment]
I'm in my bed already, I will give you some hint tomorrow. My branch is in a fork so it's more complex to add in same repo you would need to add a remote. 
Maybe do a fresh clone somewhere but of atsju/DFTFringe would be easiest. And somehow duplicate your build environment. 
You could also add me to your git if that's ok. Then I will do a branch like gr did. 
Zip must also be available in atsju/DFTFringe

Anyway, let me know and I can help. 

To be clear you want to test the build because of project modification. If you want to test the outcome of CI it's a zip containing the actual exe, not exactly same need.

**Comment by gr5** [comment]
It should be easy for Dale to look at your branch.  He can pull it into his local branch.

So if you click the red circled link (circled in image below) in this PR, it then shows you some instructions as you can see.  The first instruction circled in green creates a fresh branch based off master called "atsju-JST/windowsCI". That's a long name. I'd just call it PR15 or Julien or something.  Then the next command pulls in the correct branch from Juliens repository into the current branch.  At this point, the files in the current folder will be the ones in this PR.  To get back to the master branch just do:
git checkout master
and to return *again* to this branch do:
git checkout PR15   (or whatever you called it earlier).  If you make any edits, I believe git won't let you switch branches.  So you can delete a file that you edited and grab a fresh copy - for example if you had edited the ".pro" file and want to go back to before you edited it:
rm DFTFringe.pro    (or delete it using windows "del" or delete it from explorer)
git checkout DFTFringe.pro  (grabs fresh copy from current branch and commit)

Don't do the merge step.  That can be done from within github.  Make sure to switch back to master branch.  Otherwise you might forget you are on the PR15 branch.

Screenshot from 2023-02-23 16-22-45

**Comment by githubdoe** [comment]
George.  The likelihood of me messing up my local git repro of DFTFringe is very high.  Thus I want nothing to do with it when trying this pull request.  I don't want to then have to figure out how to undo stuff I did not intend.  Thus I'm hoping for a non-git based solution to getting the trial code.

I suppose if I have to I can create a temporary local git repro but then it will be tied to the remote with the possibility that I then get them confused and do x things wrong that also affect the remote.  I would like to break that link so that can not happen.
Then once I have confidence that the  new code will work as I desire than I will go back to the actual local git.

**Comment by gr5** [comment]
If you create a new folder (not a subfolder of your existing git repository) and then do a "git clone" process you can check out your repository or julien's entire repository fresh in that new folder.  The repository is in your working directory - the one with all the cpp files and such - in a hidden subfolder called ".git".  See for yourself. So if you just create another new folder somewhere else and do a git clone it will create a new copy of the whole repository.

If you mess things up you can go back to your previous folder where you were building DFTF.

Here are 3 ways to get Julien's code:

1) **ZIP file method -** go to his github page here:
https://github.com/atsju/DFTFringe

It should default to master branch but make sure.  Then click the green "Code" button and choose Download Zip

2) **clone (grab) his repository**
in some folder that isn't already a git repository - like your home folder for example do:

git clone https://github.com/atsju/DFTFringe.git PR15
Then enter that folder (PR15 or whatever you call it).  This checks out the master branch by default so you are good to start compiling.  You can't do a push unless he gives you "collaborator" access so you can't hurt anything.

3) **clone your own repository** in a new folder, maybe call it PR15 and then follow the directions in my previous post above to create a temporary branch and pull Julien's code in.


Because all the state information of a repository is in the .git subfolder, you can just delete the folder and the repository is off your machine.  So for example in method 2 above, just delete the PR15 folder when done because the repository is in a subfolder (called .git).



The last parameter is the name of the folder to put it all into.

**Comment by atsju** [comment]
> git checkout PR15 (or whatever you called it earlier). If you make any edits, I believe git won't let you switch branches. So you can delete a file that you edited and grab a fresh copy - for example if you had edited the ".pro" file and want to go back to before you edited it:
> rm DFTFringe.pro (or delete it using windows "del" or delete it from explorer)
> git checkout DFTFringe.pro (grabs fresh copy from current branch and commit)

The clean way to revert any unwanted change is `git reset --hard`
and to delete untracked files also `git clean -dxf`

------------

I understand Dale not wanting to clone 3 times the repo. I do it myself and can get confused. 

The easiest : 
**From the PR go to my correct branch in MY repo. This gives a direct link not to master but to the correct branch:**
image

-----------

**And the download the zip of it** 
image



Warning: this will be the correct code at the moment of the download. When I change the branch, you need to download again.

Also note that this PR basicaly includes the other 2. If it works, no need to test them.
This is why I want you to first merge the other 2 (to keep things separate) and the I make this one a clean unique change and then we merge.

**Comment by gr5** [comment]
Oops. Right.  You don't want his master branch.  You want his JST/windowsCI branch.  I had the right instructions the first time when I "circled it" (squared it? it's a rectangular circle) in green but somehow, later, thought the correct branch was julien's master branch.

**Comment by githubdoe** [comment]
It builds and runs on my machine once I added a config into the project file that I can use for my paths and opencv version.
I will not bother making that change to your .pro file.  I can make it later to the master.

I have forgotten what the next step should be.

**Comment by atsju** [comment]
Next step might be merge the other 2 open PRs.

Could you please elaborate on the 
> once I added a config into the project file that I can use for my paths and opencv version

And what you cant to add to master later ? I don't think we want to have 2 different .pro in the project for example. The whole point of this PR being draft is that we agree on how to do something that will fit for everybody and I can still change whatever you need. No hurry.

**Comment by githubdoe** [comment]
I added a section inside the Win32.  It is called it DOELEGACY.  To trigger it you uncomment
#CONFIG += DOELEGACY

There is only one .pro file.  

Here is the code.  I started to fix it with just a PATH variable but then that was not enough. 

#CONFIG += DOELEGACY
PATH = "..\\"
DOELEGACY {
    PATH = "D:\\\\"
}
win32 {
    DOELEGACY {
          CONFIG( debug, debug|release ) {
            # debug
            LIBS += D:\\qwt-6.1.5\\lib\\qwtd.dll
          } else {
            # release
            LIBS += D:\\qwt-6.1.5\\lib\\qwt.dll
          }
          INCLUDEPATH += D:\\qwt-6.1.5\\src

          #message("using win32")include

    INCLUDEPATH += D:\armadillo\armadillo-9.200.6\include

    INCLUDEPATH += D:\opencv\opencv-3.4.12\build\install\include

    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_core3412.dll
    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_highgui3412.dll
    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_imgcodecs3412.dll
    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_imgproc3412.dll
    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_features2d3412.dll
    LIBS += D:\opencv\opencv-3.4.12\build\bin\libopencv_calib3d3412.dll


    LIBS += D:\armadillo\bin\libarmadillo.dll
    LIBS += D:\lapack\build64\bin\liblapack.dll
    LIBS += D:\lapack\build64\bin\libblas.dll
    }

    !DOELEGACY {
      CONFIG( debug, debug|release ) {
        # debug
        LIBS += $$PATH\qwt-6.1.6\\lib\\qwtd.dll
      } else {
        # release
        LIBS += $$PATH\qwt-6.1.6\\lib\\qwt.dll
      }
      INCLUDEPATH += $$PATH\qwt-6.1.6\\src

      #message("using win32")include


        INCLUDEPATH += ..\build_armadillo\tmp\include

        INCLUDEPATH += ..\build_openCV\install\include\

        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_core460.dll
        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_highgui460.dll
        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_imgcodecs460.dll
        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_imgproc460.dll
        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_features2d460.dll
        LIBS += ..\build_openCV\install\x64\mingw\bin\libopencv_calib3d460.dll


        LIBS += ..\build_armadillo\libarmadillo.dll
        LIBS += ..\build_lapack\lib\liblapack.a
        LIBS += ..\build_lapack\lib\libblas.a
    }
}

**Comment by atsju** [comment]
I don't thing having a big if/else is good practice here. However I don't have better solution available so I will consider it as best solution because it's easy. 

Now I think you can merge #13 #14 
Then I will rebase this one before merging to have a clean history.
And after that you can work on sharing a corrected .pro and I can continue with my other ideas.

Note that once this is merged, you won't need to build yourself so much for testing external changes or PRs. You will be able to download the .exe and test it but without building the application yourself.

**Comment by githubdoe** [comment]
I must always be able to build it and will continue to do so.   That is the only way that it can be debugged that I know of.

**Comment by atsju** [comment]
> I must always be able to build it and will continue to do so. That is the only way that it can be debugged that I know of.

Of course you must be able to build to debug, I'm not trying to remove you from this project. I was aiming only the "testing" case where someone changes something in sources (except the .pro) and you only need to test the effects of the changes on the final application. You no longer need to checkout his branch, to build and to test like you did here. You only download the "artifact" containing the application and you test the final build.

**Comment by gr5** [comment]
nice.

**Comment by atsju** [comment]
Hi all. Feel free to merge this PR. I made no change in code, only squashed the commit to have a cleaner history.

I also compiled some of my ideas here https://github.com/users/atsju/projects/1/views/1 I might or might not implement them but I like to at least keep track of it. I tried to priorise from most important to least from my POV. If some of the ideas look really important or interesting to you do not hesitate to let me know.

Also if you have same kind of roadmap/todolist in your head or in a private place, you might want to share them in the [project](https://github.com/githubdoe/DFTFringe/projects?query=is%3Aopen) tab of github or in issues. This might lead to external developers making the changes for you (and more PRs to review).

**Comment by gr5** [comment]
> I also compiled some of my ideas here https://github.com/users/atsju/projects/1/views/1

Link doesn't work for me - I think it's protected or something by security settings.  Try logging out of github and ideally the link would still work.

**Comment by atsju** [comment]
Sorry, I didn't know projects have access rights now. 
I made it public and you should have access.

**Comment by githubdoe** [comment]
My favorite of those is to get the source into a source folder on github.  I could never figure out a way.   However I view it as busy work. 

In my opinion trying to keep DFTFringe using up to date libraries is an exercise in futility since OpenCV and Qt update at least twice a year.  Besides OpenCV there are plenty of Qt depreciated warnings as well.  All these cause much work that makes no difference to the end user and maybe helps little with code maintenance.  My philosophy is to only address issues like that when enhancing  or bug fixing a particular feature.  Giving priority to those things that actually improve the users experience.

**Comment by atsju** [comment]
Thank you for feedback Dale. I will try to understand what blocks with source folder then.

About keeping things up to date, there is a right balance between using old unmaintained versions and the lastest of everything. This balance is in general an LTS (long term support) version of dependencies. I believe you are still on LTS for OpenCV but not for QT.
My rationale behind using up to date things is that something will break at some point and it's more maintainable to have up to date versions. Examples that come to my mind are: old versions not being downloadable anymore, new features not available, (cyber)security issues, incompatibility between dependencies, ... The keyword is just maintainability, it brings no direct value to user. Also it's easier to do small steps to keep up to date than big bang update. If DFTfringe was to never be updated/modified anymore I would not bother updating anything but this is not the case.
I fully understand your point of view not wanting to use your time to change things that are already working. This is why I want to do it for you. So you can focus on what is interesting for you and I am not able to do.

File: .github/workflows/build-windows.yml [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+      - name: extract archive
+        if: steps.cache-QWT.outputs.cache-hit != 'true'
+        run: 7z x qwt-6.1.6.zip
+      - name: install aqtinstall tool
+        if: steps.cache-QWT.outputs.cache-hit != 'true'
+        run: pip install aqtinstall
+      - name: install qmake
+        if: steps.cache-QWT.outputs.cache-hit != 'true'
+        run: aqt install-qt windows desktop 5.15.2 win64_mingw81 --archives qtbase qtsvg
+      - name: qmaje
```
Comment (atsju) [inline .github/workflows/build-windows.yml:133]:
```suggestion
      - name: qmake
```

**Comment by gr5** [comment]
Okay so I'm pretty happy with this pull request.  

I tested the hell out of it including the PSI functions.  I don't usually do this much testing but you changed the version of opencv and I wanted to make sure nothing was affected.

I'm almost ready to pull it however I emailed Dale and he'd like to preserve a DFTFringe.pro file without the changes in this PR.  We can name it DFTFringe_dale.pro.  How does that sound @atsju?  Do you want to check in that file to this PR and then I'll approve this pull request?  Do you have an alternate plan?  I could add it in a separate PR.

If we add lots of .cpp files it may become a pain in the neck to keep adding them to both pro files but for now I think this is a reasonable request.  Also I think qtcreator will build just fine even if certain cpp files are not mentioned in the pro file so maybe this is less work than I think.  Also my next few changes to the code will not involve adding any .cpp or .h files (just editing existing ones).

Also I noticed many files were missing in the created zip file.  Nothing crucial but Dale never checked in lots of things including 3 color maps and a few dozen .jpg files needed for the help files.  So I'll fix that in an upcoming PR.  I'll create an issue right now.

**Comment by atsju** [comment]
This is more than reasonable. I added the original file.

> Also I noticed many files were missing in the created zip file. Nothing crucial but Dale never checked in lots of things including 3 color maps and a few dozen .jpg files needed for the help files. So I'll fix that in an upcoming PR. I'll create an issue right now.

Do not hesitate to be more precise in the issue #26 I (or other) can help too.

**I see one of the builds failed. I will look at it before merging.**

**Comment by atsju** [comment]
I don'k know what causes to fetch intel-media-va-driver_22.3.1+dfsg1-1ubuntu1_amd64.deb on http://azure.archive.ubuntu.com/ubuntu/pool/universe/i/intel-media-driver/ 
since 2023/04/21 the most approching name is  intel-media-va-driver_22.3.1+dfsg1-1ubuntu2_amd64.deb. 
To be investigated

**Review by gr5 — APPROVED** [review]
This is great.  I tested the created zip file somewhat thoroughly and approve this PR.

**Comment by gr5** [comment]
I see build-linux fails.  So when the linux build process was created, I had "focal" (v 20.04) on my computer and I believe github was building with "jammy" (v 22.04).  So now a new version of ubuntu is out and I'm guessing github is now using "kinetic kudu" (v 22.10).  I don't even know if the build uses ubuntu.  Anyway it seems reasonable to me to just change it back to explicitly using "jammy" rather than test out new library versions.  At least for now.

**Comment by atsju** [comment]
> r_22.3.1+dfsg1-1ubuntu1_amd64.deb on http://azure.archive.ubuntu.com/ubuntu/pool/universe/i/intel-media-driver/
> since 2023/04/21 the most approching name is intel-media-va-driver_22.3.1+dfsg1-1ubuntu2_amd64.deb.
> To be investigated

@gr5  no worries, This is already fixed in #27. 
ubuntu latest is still 22.04 https://github.com/actions/runner-images

**Pull request merged**